python/ray/serve/scripts.py

```diff
@@ -504,39 +505,92 @@ def shutdown(address: str, yes: bool):
+    kubernetes_format: bool,
+    output_path: Optional[str],
+    multi_app: bool,
 ):
     sys.path.insert(0, app_dir)
 
-    node: Union[ClassNode, FunctionNode] = import_attr(import_path)
-    if not isinstance(node, (ClassNode, FunctionNode)):
-        raise TypeError(
-            f"Expected '{import_path}' to be ClassNode or "
-            f"FunctionNode, but got {type(node)}."
-        )
+    def build_app_config(import_path: str, name: str = None):
+        node: Union[ClassNode, FunctionNode] = import_attr(import_path)
+        if not isinstance(node, (ClassNode, FunctionNode)):
+            raise TypeError(
+                f"Expected '{import_path}' to be ClassNode or "
+                f"FunctionNode, but got {type(node)}."
+            )
 
-    app = build_app(node)
-    schema = ServeApplicationSchema(
-        import_path=import_path,
-        runtime_env={},
-        host="0.0.0.0",
-        port=8000,
-        deployments=[deployment_to_schema(d) for d in app.deployments.values()],
-    )
+        app = build_app(node)
+        schema = ServeApplicationSchema(
+            import_path=import_path,
+            runtime_env={},
+            deployments=[
+                deployment_to_schema(d, not multi_app) for d in app.deployments.values()
+            ],
+        )
+        # If building a multi-app config, auto-generate names for each application.
+        # Also, each ServeApplicationSchema should not have host and port set, it should
+        # be set at the top level of ServeDeploySchema.
+        if multi_app:
+            schema.name = name
+            schema.route_prefix = app.ingress.route_prefix
+        else:
+            schema.host = "0.0.0.0"
+            schema.port = 8000
 
-    if kubernetes_format:
-        config = schema.kubernetes_dict(exclude_unset=True)
-    else:
-        config = schema.dict(exclude_unset=True)
+        if kubernetes_format:
+            return schema.kubernetes_dict(exclude_unset=True)
+        else:
+            return schema.dict(exclude_unset=True)
 
     config_str = (
         "# This file was generated using the `serve build` command "
         f"on Ray v{ray.__version__}.\n\n"
     )
-    config_str += yaml.dump(
-        config, Dumper=ServeBuildDumper, default_flow_style=False, sort_keys=False
-    )
+
+    if not multi_app:
+        if len(import_paths) > 1:
+            raise click.ClickException(
+                "Extra arguments detected. If you want to generate a multi-application "
+                "config, please turn on the feature flag `--multi-app`."
+            )
+
+        config_str += yaml.dump(
+            build_app_config(import_paths[0]),
+            Dumper=ServeApplicationSchemaDumper,
+            default_flow_style=False,
+            sort_keys=False,
+        )
+    else:
+        configs = []
+        for app_index, import_path in enumerate(import_paths):
+            configs.append(build_app_config(import_path, f"app{app_index + 1}"))
```

Comment:
If users enable `kubernetes_format` and they request a multi-app config, what's the behavior right now? It looks like all the internal Serve configs will have a K8s format, but the outer `ServeDeploySchema` will still use snake-case. Once we integrate the new schema into the K8s config, it probably won't be snake-cased. Can we either

1. Raise an error if a user requests a `multi-app` config and `kubernetes_format`,
2. raise a warning if a user requests a `multi-app` config and `kubernetes_format`,
3. or come up with the actual design of the future KubeRay schema and output that if the user requests both?

Reply:
Thanks for catching that! I think we should go with 1 or 2 for now, since we haven't thought about the design for KubeRay yet. I'm leaning more towards 1, WDYT?

Reply:
I think 1 makes sense!

Reply:
Added a check!

Reply:
good catch!